text/0000-never-patterns.md

```diff
@@ -0,0 +1,250 @@
+- Feature Name: `never_patterns`
+- Start Date: 2024-10-27
+- RFC PR: [rust-lang/rfcs#3719](https://github.com/rust-lang/rfcs/pull/3719)
+- Rust Issue: [rust-lang/rust#118155](https://github.com/rust-lang/rust/issues/118155)
+
+# Summary
+[summary]: #summary
+
+A `!` pattern indicates a type with no valid values. It is used to indicate an impossible case when matching an empty type in unsafe code.
+
+```rust
+enum Void {}
+unsafe fn empty_tup<T>(tup: *const (T, Void)) -> ! {
+    unsafe {
+        match *tup { ! }
+    }
+}
+```
+
+Note: this RFC is purely about a new pattern syntax. It does not propose changes to exhaustiveness checking or to the operational semantics of uninhabited types.
+
+
+# Motivation
+[motivation]: #motivation
+
+Rust's unsafe semantics are access-based: the validity of data only matters when reading from/writing to a place. When pattern-matching, patterns only access the data (discriminants/values) they need to choose the right arm.
```

Comment:
Rust's unsafe semantics are actually mostly *not* access-based. Constructing an invalid value is UB.

Only behind raw pointers and inside union fields do we use access-based semantics. So the text here is quite misleading as-is.

Reply:
Niko's [blog post from year 2018](https://smallcultfollowing.com/babysteps/blog/2018/08/13/never-patterns-exhaustive-matching-and-uninhabited-types-oh-my/), linked in the RFC text, says that

> While the details of our model around unsafe code are still being worked out (in part by this post!), there is a general consensus that we want an “access-based” model

...mentioning about `&!` being possibly a valid value. Just to clarify, I presume this is now stale information and having a type `&!` is considered UB? Also, the following code is presented:

```rust
unsafe {
  let x: Uninit<(u32, T)> = Uninit { uninit: () };
  x.value.0 = 22; // initialize first part of the tuple
  ...
  match x.value {
    (v, _) => {
      // access only first part of the tuple
    }
  }
  ...
}
```

Is that accepted by modern standards? Seems a bit sketchy that the union field is already accessed in the match expression, so the fields are not matched against _inside_ the match. By extension, one should be able to do the following, which, by modern standards, is surely UB, right?

```rust
unsafe {
  let x: Uninit<(u32, T)> = Uninit { uninit: () };
  x.value.0 = 22; // initialize first part of the tuple
  ...
  let tmp = x.value; // move from x.value!
  match tmp {
    (v, _) => {
      // access only first part of the tuple
    }
  }
  ...
}
```

Reply:
A match expression works on *places*, not values. So the two examples you show are not equivalent! The first never turns `x.value` into a value, it only ever accesses its first field.

Reply:
But values don't exist in memory, they only exist while reading or writing, right? And matches don't access compound values wholesale, they only inspect memory piecewise. In particular I can mess up some bytes of a place and still match on it if the match only inspects the non-messed-up bytes.

Is that not "access-based"? Am I incorrect in the above understanding?

Reply:
In an access-based semantics, it is generally expected that e.g.
```rust
let b: bool = transmute(13u8);
```
is fine if we never "access" `b`. So I think it is misleading to call Rust's semantics access-based.

But what you say about places is correct.

Reply:
I see. Is there a term of art for the particular flavor of "validity only matters when reading/writing" that rust has?

Otherwise I don't can do without this general statement. Would this be ok: "The semantics of pattern-matching around unsafe values are access-based: patterns only access the data (discriminants/values) they need to choose the right arm. The accessed data must be valid, but the rest may be invalid or uninitialized"?

Reply:
I'm not aware of a specific term. I would summarize it as "enforcing language invariants on every typed copy".

> "The semantics of pattern-matching around unsafe values are access-based: patterns only access the data (discriminants/values) they need to choose the right arm. The accessed data must be valid, but the rest may be invalid or uninitialized"

"unsafe value" isn't a standard term. I've used "unsafe places" in the past but it is not standard either; the RFC should define it if it wants to use it.

I would say something like: Rust's pattern match semantics are access-based. Only the parts of the underlying value representation that actually matter for the pattern have to be valid.